runtime/platform.c

```diff
@@ -130,6 +130,178 @@ void caml_plat_cond_free(caml_plat_cond* cond)
   cond->mutex=0;
 }
 
+/* Barriers */
+
+#ifndef CAML_PLAT_FUTEX_FALLBACK
+#  if defined(_WIN32)
+#    include <synchapi.h>
+#    define CAML_PLAT_FUTEX_WAIT(ftx, undesired)                \
+  WaitOnAddress(ftx, &undesired, sizeof(undesired), INFINITE)
+#    define CAML_PLAT_FUTEX_WAKE(ftx) \
+  WakeByAddressAll(ftx)
+
+#  elif defined(__linux__)
+#    include <linux/futex.h>
+#    include <sys/syscall.h>
+#    define CAML_PLAT_FUTEX_WAIT(ftx, undesired)    \
+  syscall(SYS_futex, ftx, FUTEX_WAIT_PRIVATE,       \
+          /* expected */ undesired,                 \
+          /* timeout */ NULL,                       \
+          /* ignored */ NULL, 0)
+#    define CAML_PLAT_FUTEX_WAKE(ftx)           \
+  syscall(SYS_futex, ftx, FUTEX_WAKE_PRIVATE,   \
+          /* count */ INT_MAX,                  \
+          /* timeout */ NULL,                   \
+          /* ignored */ NULL, 0)
+
+/* #  elif defined(__APPLE__)
+   macOS has __ulock_wait which is used in implementations of libc++,
+   (e.g. by LLVM) but the API is private and unstable. */
+
+#  elif defined(__FreeBSD__)
+#    include <sys/umtx.h>
+#    include <sys/syscall.h>
+#    define CAML_PLAT_FUTEX_WAIT(ftx, undesired)        \
+  syscall(SYS__umtx_op, ftx, UMTX_OP_WAIT_UINT_PRIVATE, \
```

Comment:
On FreeBSD, you should not invoke system calls directly but use the libc wrapper, in this case `_umtx_op()`. And then you don't need to include `<sys/syscall.h>` anymore.